BUG: lib: Fix a problem with vectorize with default parameters.

WarrenWeckesser opened the pull request:
When `otypes` is given to `vectorize` and then the instance is
called, it creates a ufunc by calling numpy.core.umath.frompyfunc.
The number of arguments given to this ufunc is set to the number
of arguments in the call of the vectorize instance.  This ufunc
is cached, so frompyfunc does not have to be called on the next
call. The problem is that, if the function being wrapped has
parameters with default values, the number of arguments passed
to the vectorize instance can change, and when that happens, a
new ufunc must be created by calling frompyfunc with the correct
number of arguments.

This commit changes the cache of the ufunc from a simple attribute
that holds the most recent ufunc to a dictionary whose keys are
the number of arguments in the call.  The cache is only used when
the vectorized function is called with only positional arguments
and there are no excluded arguments.  If keywords are used, the
number of arguments is no longer sufficient to uniquely identify a
previously created ufunc.

Closes gh-16120.

WarrenWeckesser committed:
BUG: lib: Fix a problem with vectorize with default parameters.

When `otypes` is given to `vectorize` and then the instance is
called, it creates a ufunc by calling numpy.core.umath.frompyfunc.
The number of arguments given to this ufunc is set to the number
of arguments in the call of the vectorize instance.  This ufunc
is cached, so frompyfunc does not have to be called on the next
call. The problem is that, if the function being wrapped has
parameters with default values, the number of arguments passed
to the vectorize instance can change, and when that happens, a
new ufunc must be created by calling frompyfunc with the correct
number of arguments.

This commit changes the cache of the ufunc from a simple attribute
that holds the most recent ufunc to a dictionary whose keys are
the number of arguments in the call.  The cache is only used when
the vectorized function is called with only positional arguments
and there are no excluded arguments.  If keywords are used, the
number of arguments is no longer sufficient to uniquely identify a
previously created ufunc.

Closes gh-16120.

WarrenWeckesser commented:
It would be possible to make the caching smarter, so the ufunc is also cached when keyword arguments are used, but a few experiments that I tried (e.g. make the cache key a tuple that includes the keyword names) didn't show any gain in performance.   Unless someone discovers a serious performance regression, I'll leave optimization as a possible task for future research.  For now, this PR fixes the bug.

seberg commented:
One squirm I have with this, is that ufuncs do not support variadic arguments (or kwargs). But then, I suppose that people who really want a true `ufunc` will have to use `np.frompyfunc` directly (which at some point the documentation should maybe point out, `vectorize` is such a nice name, but a proper ufunc has its advantages!).

WarrenWeckesser commented:
I don't understand your concern.  `vectorize` supports keyword arguments.  This PR does not change the API at all.

seberg commented:
Yeah, you are right.

seberg reviewed (approved):
LGTM, tests could maybe be more concise, but honestly, they seem to cover the important things. Since we only support 32 arguments in ufuncs anway, there is no reason to be curious about the theoretical maximum size of that dictionary ;).

Will probably just merge next time I look at this, unless someone beats me to it.

WarrenWeckesser committed:
MAINT: lib: A few changes in some vectorize tests.

Added a note to some of the tests that the order of the calls is an
important part of the test, because the code is testing the caching
of the ufuncs.

Also reuse a couple functions in the vectorize tests by defining them at
the module level.

WarrenWeckesser commented:
> tests could maybe be more concise,

I pushed a commit that moves the common functions used in the new tests to the module level so they can be reused.  I also added comments about the order of the function calls being important for the test, so a year from now someone doesn't "simplify" the tests.

If you have specific changes that you would like to see, let me know.  I don't mind iterating, even for small changes.

seberg commented:
Not sure the `foo` name is good, but on the other hand there is no obvious name I guess. I think the whole differnent order and kwarg vs. arg tries just read repetitive...

mattip commented:
Thanks @WarrenWeckesser
